typing/path.mli

```diff
@@ -24,9 +24,40 @@ val same: t -> t -> bool
 val compare: t -> t -> int
 val find_free_opt: Ident.t list -> t -> Ident.t option
 val exists_free: Ident.t list -> t -> bool
+
 val scope: t -> int
+        (** @raise [No_scope] if the path contains an unscoped identifier. *)
+
 val flatten : t -> [ `Contains_apply | `Ok of Ident.t * string list ]
 
+val subst : (Ident.t * Ident.t) list -> t -> t
+        (** Substitute identifiers i_1 for i_1', ... i_n for i_n' from the list
+            of the form [[(i_1, i_1'); ...; (i_n; i_n')]].
+
+            The returned value will be physically equal to the input if no
+            substitutions are made. *)
+
+val unsubst : (Ident.t * Ident.t) list -> t -> t
```

Comment:
The encoding of `id_pairs` was a cheap and efficient way to represent 2 substitutions: a 'to environment' substitution implemented by `subst` and a 'to local type' substitution implemented by `unsubst`.

Consider for example
```ocaml
module type TYPE1 = sig type 'a t end
module type S = sig
  type t
  type 'a f = {T : TYPE} -> 'a T.t -> 'a T.t * t
end
val x : {M : S} -> M.t M.f
val y : {N : S} -> {T2 : TYPE} -> N.t T2.t -> N.t T2.t * N.t

let z b = if b then x else y
```
When typechecking `z`, at the point where we will expand `M.f` we use
* `M` -> `M/4` (mapping `M` to the identifier `M/4` in the environment)
* `M` <- `M/4` (mapping `M/4` in the environment back to the local `M`)
* `T` -> `T/1` (mapping `T` to some fresh unscoped identifier `T/1`)
* `T/1` <- `T/1` (override the implicit `T` <- `T/1` mapping to ensure that `T` is not restored)

this is encoded as `[(T/1, T/1); (T, T/1); (M, M/4)]`, using the `subst` function for the -> arrows and the `unsubst` function for the <- arrows.

The steps are roughly:
* substitute in the 'head' of `M.t M.f` to get `M.t M/4.f`
* expand `M.t M/4.f` in the environment to get `'a = M.t`, `{T : TYPE} -> 'a T.t -> 'a T.t * M/4.t`
* copy this type to get `{T/1 : TYPE} -> M.t T/1.t -> unit T/1.t * M.t`.

To then unify the RHS of the `{_} -> _` arrows, we carry the substitutions `M` -> `M/5`, `N` -> `M/5`, `T/1` -> `T/3`, `T2` -> `T/3` instead of copying to substitute. This leaves us unifying essentially
```
x {M/5} {T/3} : M/5.t T/3.t -> M/5.t T/3.t * M/5.t
y {M/5} {T/3} : M/5.t T/3.t -> M/5.t T/3.t * M/5.t
```

---

In the normal case of unification and related functions, the substitutions are conceptually very simple, being a list of (local ident, environment ident) pairs. Perhaps the extra complexity around copying is too confusing though; could you can suggest a nicer way to track the substitutions we need?

It's been quite a while since I made this PR, but I think my concern with `copy` was that the path should be correct when passing to `proper_abbrevs`, otherwise the same expansion may be used for an identical path even when it's subject to different substitutions. If we can rule this out, then copy can get away with using only `unsubst`.

Reply:
You shouldn't ever need the "subst" direction. `proper_abbrevs` could well require some care, but I would expect it to be workable without "subst". Could you elaborate on the potential issue if you don't use "subst"?

Reply:
> Could you elaborate on the potential issue if you don't use "unsubst"?

One such example is `update_level`.

```ocaml
module type S = sig
  module F : functor (M : sig type t end) -> sig type t = A of M.t end
end
module Int = struct type t = int end
let f =
  let module Int2 = Int in
  let f {M : S} (x : Int2.t) : M.F(Int2).t = A x in
  f;;
```
```ocaml
val f : {M : S} -> int -> M.F(Int).t
```

Without `unsubst`, `update_level` will expand this to `val f : {M : S} -> int -> M/1.F(Int).t`, where `M/1` is some ephemeral identifier used to place `M` in the environment at the correct level.

Reply:
Thanks that is a really useful example.

Reply:
I'm a bit slow to react (sorry) but I concur, I had missed that situation so thanks for spotting (and sharing it).
I'll keep it in mind while going over the other uses of `unsubst`.

FTR: offline (well, off this line) @lpw25 came up with a variation of this example which more directly illustrates why one truly needs to add `M` in the environment at a reasonable level.
It is the following:
```ocaml
module type S = sig
  type t
  type 'a s = t
end
let f =
  let module Int2 = Int in
  let f {M : S} : Int2.t M.s = assert false in
  f;;
val f : {M : S} -> M.t
```


If you haven't done so already, could you make sure that these examples are added to the testsuite?

Reply:
I've updated the testsuite with an equivalent test (and some related ones). Closing this thread as resolved.